### PR TITLE
render_opengl: Implement anisotropic filtering

### PR DESCRIPTION
**Pull request by adityaruplaha**
Optional feature. Enabled by default.
Throws a warning if is enabled but not supported.

# Screenshots
capture

# Game Screenshots
disabled
^ Disabled
enabled
^ Enabled

I think Kukui's cap looks a little bit nicer (less pixelated). Both were taken at 2x Native with no significant difference in speed.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3589)

### CODE REVIEW

**Comment by legoj15** [comment]
Not to bike-shed/nit, but I feel like "look nicer" is a completely suggestive term, and that anisotropy should have the 2x ,4x, 8x,16x options. A single checkbox reminds me of Dophin's "Force Texture Filtering". Just a user's perspective.

File: src/video_core/renderer_opengl/gl_rasterizer.cpp [hunk 1]
```diff
@@ -1197,6 +1198,18 @@ void RasterizerOpenGL::SamplerInfo::SyncWithConfig(
             glSamplerParameterfv(s, GL_TEXTURE_BORDER_COLOR, gl_color.data());
         }
     }
+
+    // use anisotropic filtering
+    if (VideoCore::g_use_anisotropic_filtering) {
+        if (GL_ARB_texture_filter_anisotropic) {
+            GLfloat anisoMax;
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_rasterizer.cpp]: `aniso_max`

File: src/video_core/renderer_opengl/gl_rasterizer.cpp [hunk 2]
```diff
@@ -1197,6 +1198,18 @@ void RasterizerOpenGL::SamplerInfo::SyncWithConfig(
             glSamplerParameterfv(s, GL_TEXTURE_BORDER_COLOR, gl_color.data());
         }
     }
+
+    // use anisotropic filtering
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_rasterizer.cpp]: The variable name already explains what this block of code does.

File: src/video_core/renderer_opengl/gl_rasterizer.cpp [hunk 3]
```diff
@@ -1197,6 +1198,18 @@ void RasterizerOpenGL::SamplerInfo::SyncWithConfig(
...
+    if (VideoCore::g_use_anisotropic_filtering) {
+        if (GL_ARB_texture_filter_anisotropic) {
+            GLfloat anisoMax;
+            glGetFloatv(GL_MAX_TEXTURE_MAX_ANISOTROPY_EXT, &anisoMax);
+            glSamplerParameterf(s, GL_TEXTURE_MAX_ANISOTROPY_EXT, anisoMax);
+        } else {
+            // Not a fatal problem , so only a warning should do here.
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_rasterizer.cpp]: `,` should be against `problem`

**Comment by wwylele** [comment]
Despite what you said on discord, I... honestly still can't see how one image is better than the other. I suspect that you assumed this thing should improve the graphics first, which forced you to feel that it actually looks better.

**Comment by tabnk** [comment]
Performance Drop Significantly

Super Mario 3D Land:
Before **163 fps**
After **122fps**

**Comment by adityaruplaha** [comment]
Anisotropic filtering is quite expensive, you know...

**Comment by legoj15** [comment]
The only thing that currently puzzles me right now is that (atleast for me from what i've seen from 90% of my games) all textures seem to be at their highest mip-level on nightly and canary, which means that at all angles and distances the texture is at it's full resolution. From what I know about anisotropic filtering in video games, the levels of anisotropy are supposed to increase the texture fidelity at indirect angles and far distances, but since the textures are already at their maximum resolution, what is this doing? I don't see any difference at all in the currently provided screenshots.

I may be going off on a tangent here but I must ask these questions:

- Does Citra currently support mip-maps?
- What level of anisotropy is being set when the box is checked? (1x, 2x, 4x, 8x, 16x)

PS: I still feel that a single checkbox is more akin to forcing filtering on all textures, whether they're specified to be pixelated by the game or not. I hate to keep making parallels to Dolphin but as a user of it, I would expect the checkbox to make textures that look like this (unfiltered):
image
Look like this (filtered):
image

I apologize if this is just a wall of text and images, but as somebody who wants the best user experience from Citra, I'm just a bit confused as to what this setting currently does.

(also to answer @adityaruplaha 's question, I don't think that a drop down menu would be too complex, but then again, I'm not the UX expert of this project)

**Comment by adityaruplaha** [comment]
This PR is useless currently, because mipmapping isnt implemented. I'll suggest you look the conceptual part of ThinMatrix's LWJGL Tutorial 41. You can find it on YouTube.

**Comment by adityaruplaha** [comment]
Feel free to reuse this code whenever required.

**Pull request closed without merging**